### ISSUE TITLE
Power management support

### ISSUE THREAD

**Issue by Tibladar** [pwr-Solaar/Solaar#1949]
**Information**
- Solaar version: 1.1.8
- Distribution: Ubuntu 22.04

Currently there's no support for setting the power management setting for headsets (at least for the Logitech G535)
It controls when the headset turns itself off after loosing the connection to the usb receiver (usually after turning off the pc)
![grafik](https://user-images.githubusercontent.com/52620063/210227380-cca26a55-bea8-426a-9145-732119c96cfe.png)
![grafik](https://user-images.githubusercontent.com/52620063/210227404-79183f75-ca8e-4f5e-951c-3fc38327293c.png)

Here are two wireshark captures using the G Hub software under Windows:
- Setting power management to 1 minute: [1minute.pcapng.gz](https://github.com/pwr-Solaar/Solaar/files/10331327/1minute.pcapng.gz)
-  Setting power management to never, cycling through all settings and setting it to never: [pm_loop.pcapng.gz](https://github.com/pwr-Solaar/Solaar/files/10331336/pm_loop.pcapng.gz)

If something else is needed please let me know.

**Comment by pfps** [comment]
This looks like something that Solaar might be able to do.   The relevant part of the communication appears to be using an HID++ message, but one that I do not have documentation for.  If you capture interactions for all the values it might be possible to write code for a Solaar setting for this.

**Comment by pfps** [comment]
Ooops I was looking at only the first capture.  I see you provided them all.

**Comment by pfps** [comment]
If you are using this headset under Linux is would be useful to see what values are possible for this function.    Solaar has a tool to send HID++ commands to devices which is available if you clone Solaar from this repository.

Find out the hidraw node for the headset.  Solaar shows this in the detailed information pane - click on the information or lightbulb button.

From the Solaar repository clone run `./tools/hidconsole /dev/hidrawN` for the appropriate N.  The send commands like
```
11 ff 05 2b XX 000000000000000000000000000000
```
for XX ranging from 00 up to 1e and beyond until you get uniform responses that have 05 in the fourth byte instead of the third.   (These are error responses - normal responses will have the 05 in the third byte.)  

Post the result.

**Comment by pfps** [comment]
That's helpful.  It looks as if all values from 0 to 47 work fine.  Could you try to do a search to find the highest working value?  First try xFF to see if all values work.  Other potential highest working values are x64 (100) and x3c (60).

**Comment by pfps** [comment]
PR #1952 has the first pass at a setting for ADC power management.  Please run it and see whether it works.  Currently it allows any value between 1 and 255 plus 0 for never.

To clone and use Solar from its GitHub repository
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1952, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1952/head:pull_1952
git checkout pull_1952
```
To download a new version of the pull request, fetch it and then set your pull branch to the new fetch, as in:
```
git checkout pull_1952
git fetch origin pull/1952/head
git reset --hard FETCH_HEAD
```

**Comment by pfps** [comment]
So it looks as if all 256 values are accepted.   Can you try a value that is not in the G Hub list and see whether the headset turns off as expected?  Try something small, like 4 or 6.

**Comment by pfps** [comment]
Also see what `solaar show` produces after you try the non-list value

**Comment by pfps** [comment]
Can you try with the headset itself active?   Solaar is seeing error messages that I think are caused by the headset being in power-saving mode.

**Comment by pfps** [comment]
OK.  That looks good.  Try with a power management value of 7 (or something else not on the official list).

**Comment by pfps** [comment]
Hi.  It looks like the power management timer is in non-volatile memory.  It may be that this memory can wear out.  With the PR here Solaar is writing to that memory a lot.  I'll put together an enhancement to Solaar to prevent this, but I suggest not using the PR for now.

**Comment by pfps** [comment]
PR #1964 only writes the setting if it needs to be changed and will reduce the write load on the device.  Please try it out.

To clone and use Solar from its GitHub repository
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1964, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1964/head:pull_1964
git checkout pull_1964
```
To download a new version of the pull request, fetch it and then set your pull branch to the new fetch, as in:
```
git checkout pull_1964
git fetch origin pull/1964/head
git reset --hard FETCH_HEAD
```

**Comment by pfps** [comment]
The debugging messages are somewhat misleading so don't worry about that.  The problem is that the value is still being written.  I'll fix up the debugging messages to better show what is going on.

**Comment by pfps** [comment]
OK, there is a new version of the PR with better debugging messages.  Please try it out.

**Comment by pfps** [comment]
Aah, the 'current value' is not being stripped of all the trailing bytes because of some aspect of the setting.

**Comment by pfps** [comment]
OK, there is a new version of the PR that first interprets the bytes that were read from the device and only then does the comparison.

**Comment by pfps** [comment]
Yeah, silly me.  I didn't add the changed file before committing.  I did a force push so that didn't complain, that would have reminded me of my error.

**Comment by pfps** [comment]
There should be a new version on github.  The main changes are in lib/logitech_receiver/settings.py around line 1185.  Look for
```
        current_value = self.validate_read(current_value) if current_value is not None else None
```
if you need to check that you have the most-recent version.

**Comment by pfps** [comment]
OK. That looks correct.  The power management value is read from the device and, as it is the same as the desired value, nothing is written.
